Refactor explicit FLS usage in the runtime

janvorli opened the pull request:
This change replaces explicit FLS usage for thread shutdown notification in the runtime into a
general thread shutdown callback registration mechanism.
The implementation using the FLS on Windows is moved to the PalRedhawkWinMin.cpp.
On Unix, it is implemented using TLS, since there are no fibers.

janvorli commented:
@jkotas, @smosier Can you take a look please?

janvorli commented:
Please note that setting the FLS value to the current thread pointer is now done before initializing the thread. I believe the change doesn't introduce any functional problem.

janvorli commented:
@smosier I was trying to preserve the saveguard checks in the ThreadStore::AttachCurrentThread, but  if we don't care about failing fast in case a thread is registered with a fiber and ThreadStore::AttachCurrentThread is called on another fiber and we are ok just doing nothing in that case, I could simplify it further by removing the isFirstAttach parameter of the PalAttachThread and returning from the ThreadStore::AttachCurrentThread without doing anything when the pAttachingThread->IsInitialized() is true.

janvorli commented:
Implements #640

jkotas commented on src/Native/Runtime/PalRedhawk.h:
```diff
@@ -788,6 +788,10 @@ REDHAWK_PALIMPORT UInt32_BOOL REDHAWK_PALAPI PalAllocateThunksFromTemplate(_In_
 
 REDHAWK_PALIMPORT UInt32 REDHAWK_PALAPI PalCompatibleWaitAny(UInt32_BOOL alertable, UInt32 timeout, UInt32 count, HANDLE* pHandles, UInt32_BOOL allowReentrantWait);
 
+REDHAWK_PALIMPORT bool REDHAWK_PALAPI PalRegisterThreadShutdownCallback(ThreadShutdownCallback callback);
```
I do not look at this PAL as independent piece. I think it would be ok to just call extern C method with given name from the PAL, and not have this registration.

jkotas commented on src/Native/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -810,31 +810,87 @@ extern "C" void RaiseFailFastException(PEXCEPTION_RECORD arg1, PCONTEXT arg2, UI
+//  isFirstAttach - true if this is the first attach for the given thread
+// Return:
+//  true if the thread was not attached to PAL, false if it was
+extern "C" bool PalAttachThread(void* thread, bool isFirstAttach)
```
The `isFirstAttach` argument looks like unnecessary complexity as you have mentioned in the comments ... it would be nice to get rid of it.

jkotas commented:
LGTM otherwise

smosier commented:
The original point of this fiber business was move the thread detach code to a place where the Win32 loader lock was not held.  It's been too long for me to remember much more detail, but I will dig into it and get back to you.  Do we have to worry about anything like the loader lock on unix?

janvorli commented:
@smosier There is not such issue on Unix.

janvorli commented:
@jkotas, @smosier  I have changed few things:
1) Removed the callback registration as you've suggested
2) Placed the registration code to PalInit
3) Removed the constructor attribute from the PalInit and add call to it to InitDLL instead so that it is called both on Windows and Unix

jkotas commented:
Link error on Ubuntu

janvorli commented:
@jkotas, the PalInit in PalRedhawkUnix.cpp was missing REDHAWK_PALEXPORT and REDHAWK_PALAPI annotation.

manu-silicon commented on src/Native/Runtime/threadstore.cpp:
```diff
@@ -177,7 +155,7 @@ void ThreadStore::AttachCurrentThread()
+void ThreadStore::DetachCurrentThread()
 {
     //
     // Note: we call this when each *fiber* is destroyed, because we receive that notification outside
```
Shouldn't the comment be updated to remove the reference to fiber?

janvorli commented:
@jkotas, @smosier I have removed the isFirstAttach parameter from the PalAttachThread and submitted it as a separate commit so that you can compare it to the previous version.

jkotas commented on src/Native/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -313,6 +333,65 @@ REDHAWK_PALEXPORT bool REDHAWK_PALAPI PalHasCapability(PalCapability capability)
+
+    if (attachedThread != NULL)
+    {
+        ASSERT_UNCONDITIONALLY("PalAttachThread called multiple times for the same thread");
```
Nit - Regular assert may be more appropriate here.

jkotas commented:
LGTM

smosier commented on src/Native/Runtime/startup.cpp:
```diff
@@ -39,13 +39,18 @@ UInt32 _fls_index = FLS_OUT_OF_INDEXES;
 
 bool InitDLL(HANDLE hPalInstance)
 {
+    if (!PalInit())
```
Calling PalInit here will mean that it gets called twice in the PN runtime, which calls PalInit from DllMain.  The reason for PN to do it this way was that we considered the runtime to be three separate parts:  unmanaged (C++), managed (C#), and PAL.  DllMain was responsible for initializing all 3 in the proper order.  Putting the PalInit here will conflate the C++ and PAL pieces together and require some corresponding PN-side changes.  Could we call PalInit from __initialize_runtime instead?

smosier commented:
In general, I'd feel more comfortable about the FLS changes if you were to at least prototype the corresponding changes in the "app PAL" used by PN.  This would help us understand if there are any unique issues there which might affect the form of these changes.

smosier commented on src/Native/Runtime/windows/PalRedhawkMinWin.cpp:
```diff
@@ -67,6 +92,12 @@ REDHAWK_PALEXPORT bool REDHAWK_PALAPI PalInit()
     if (!PalQueryProcessorTopology())
         return false;
 
+    g_flsIndex = FlsAlloc(FiberDetachCallback);
```
Please add a comment about what we're doing with FLS here.  Basically something along the lines of "we use fiber detach callbacks to run our thread shutdown code because the fiber detach callback is made without the OS loader lock" is what I'd like to see.

smosier commented on src/Native/Runtime/startup.cpp:
```diff
@@ -214,13 +215,10 @@ void DllThreadDetach()
 }
 
-void __stdcall FiberDetach(void* lpFlsData)
+void __stdcall ThreadShutdown()
```
Nit:  this name is fairly generic.  And, given that this is a contract with the PAL, I wonder if we should have a less-generic name.  My suggestion would be `RuntimeThreadShutdown`, although I must admit that even this is fairly generic.

smosier commented on src/Native/Runtime/windows/PalRedhawkMinWin.cpp:
```diff
@@ -58,6 +61,28 @@ bool InitializeSystemInfo()
+// in the FLS value, others have NULL there.
+void __stdcall FiberDetachCallback(void* lpFlsData)
+{
+    UNREFERENCED_PARAMETER(lpFlsData);
```
why is this needed?  lpFlsData is used below...

janvorli replied:
This is not needed anymore. There used to be just the assert before using it, so that's why it was there to prevent warnings in release builds.

smosier commented on src/Native/Runtime/windows/PalRedhawkMinWin.cpp:
```diff
@@ -58,6 +61,28 @@ bool InitializeSystemInfo()
+    }
+
+    // The thread is shutting down
+    ThreadShutdown();
```
I think it might be valuable if we were to pass lpFlsData down to ThreadShutdown so that ThreadShutdown can assert that ThreadStore::GetCurrentThread() matches the lpFlsData value.  Is this possible also on unix, or does it not make any sense there?

smosier commented on src/Native/Runtime/windows/PalRedhawkMinWin.cpp:
```diff
@@ -58,6 +61,28 @@ bool InitializeSystemInfo()
 
+void __stdcall ThreadShutdown();
+
+// Callback that is called when a fiber is detached from a thread
```
nit: to me, this line of the comment seems rather redundant on a method called `FiberDetachCallback`. Can you either elaborate some more on what's going on (i.e. how this is just the mechanism we use to perform thread detach outside of the loader lock) or just eliminate it altogether.

janvorli commented:
@smosier I've reflected all your feedback. I am going to try to create the prototype change that you've asked for.

smosier commented:
LGTM
